pcl_ros/include/pcl_ros/pcl_node.hpp

```diff
@@ -41,65 +41,103 @@
 
 **/
 
-#ifndef PCL_ROS__PCL_NODELET_HPP_
-#define PCL_ROS__PCL_NODELET_HPP_
+#ifndef PCL_ROS__PCL_NODE_HPP_
+#define PCL_ROS__PCL_NODE_HPP_
 
-#include <sensor_msgs/PointCloud2.h>
-// PCL includes
-#include <pcl_msgs/PointIndices.h>
-#include <pcl_msgs/ModelCoefficients.h>
-#include <pcl/pcl_base.h>
+#include <rclcpp/rclcpp.hpp>
 #include <pcl/point_types.h>
+#include <pcl_msgs/msg/point_indices.hpp>
+#include <pcl_msgs/msg/model_coefficients.hpp>
 #include <pcl_conversions/pcl_conversions.h>
-// ROS Nodelet includes
-#include <nodelet_topic_tools/nodelet_lazy.h>
+#include <tf2_ros/buffer.h>
+#include <tf2_ros/transform_listener.h>
 #include <message_filters/subscriber.h>
 #include <message_filters/synchronizer.h>
 #include <message_filters/sync_policies/exact_time.h>
 #include <message_filters/sync_policies/approximate_time.h>
-
-// Include TF
-#include <tf/transform_listener.h>
-
-// STL
+#include <sensor_msgs/msg/point_cloud2.hpp>
+#include <memory>
 #include <string>
-
-#include "pcl_ros/point_cloud.hpp"
+#include <vector>
 
 using pcl_conversions::fromPCL;
 
 namespace pcl_ros
 {
-////////////////////////////////////////////////////////////////////////////////////////////
-////////////////////////////////////////////////////////////////////////////////////////////
-////////////////////////////////////////////////////////////////////////////////////////////
-/** \brief @b PCLNodelet represents the base PCL Nodelet class. All PCL nodelets should
-  * inherit from this class.
-  **/
-class PCLNodelet : public nodelet_topic_tools::NodeletLazy
+/** \brief @b PCLNode represents the base PCL Node class.
+  * All PCL node should inherit from this class.
+  */
+class PCLNode : public rclcpp::Node
 {
 public:
-  typedef sensor_msgs::PointCloud2 PointCloud2;
+  typedef sensor_msgs::msg::PointCloud2 PointCloud2;
 
   typedef pcl::PointCloud<pcl::PointXYZ> PointCloud;
-  typedef boost::shared_ptr<PointCloud> PointCloudPtr;
-  typedef boost::shared_ptr<const PointCloud> PointCloudConstPtr;
+  typedef PointCloud::Ptr PointCloudPtr;
+  typedef PointCloud::ConstPtr PointCloudConstPtr;
+
+  typedef pcl_msgs::msg::PointIndices PointIndices;
+  typedef PointIndices::SharedPtr PointIndicesPtr;
+  typedef PointIndices::ConstSharedPtr PointIndicesConstPtr;
+
+  typedef pcl_msgs::msg::ModelCoefficients ModelCoefficients;
+  typedef ModelCoefficients::SharedPtr ModelCoefficientsPtr;
+  typedef ModelCoefficients::ConstSharedPtr ModelCoefficientsConstPtr;
+
+  /** \brief Constructor. */
+  PCLNode(std::string node_name, const rclcpp::NodeOptions & options)
+  : rclcpp::Node(node_name, options),
+    tf_buffer_(this->get_clock()),
+    tf_listener_(tf_buffer_)
+  {
+    {
+      rcl_interfaces::msg::ParameterDescriptor desc;
```

Comment:
This is a really unusual way to declare parameters, can you explain why this is necessary? Why read only?

Reply:
To be honest, the reason is because I just took the implementation from pr #255 as a straight copy and didn't think much about it. I'll have to read up on how to declare parameters in ROS2 to do this right.